src/main/java/org/springframework/integration/kafka/dsl/KafkaMessageDrivenChannelAdapterSpec.java

```diff
@@ -152,6 +156,11 @@ public S filterInRetry(boolean filterInRetry) {
 		return _this();
 	}
 
+	@Override
+	public Map<Object, String> getComponentsToRegister() {
+		return Collections.singletonMap(this.container, getId());
```

Comment:
This `getId()` is used to register a `KafkaMessageDrivenChannelAdapter` from this `KafkaMessageDrivenChannelAdapterSpec`. We can't re-use the same `id` for the `container`.

More over that might not be what we expect to do with an external container anyway.

We don't do such a registration in the JMS DSL, although we do in the AMQP...

Well, I won't mind to do this as you propose, but let's consider a `AmqpInboundChannelAdapterSMLCSpec` variant where we expose `configureContainer()` callback to let end-user to inject a desired `id`, even if a `container` is an external object. 

The JMS inconsistency we can reconsider later or on demand.

Reply:
Yes; originally I had `getId() + ".container"` but getId() can be `null`.

I guess we could use `getId() == null ? null : getId() + ".container"` ?

Reply:
I think we had such an approach before, but then decided to go with the `null` and auto-generated id by default and let end-user to provide a desired id via `configureContainer()` in the AMQP DSL.

WDYT?

Reply:
I don't think `configureContainer` is possible in this mode - the factory methods take an abstract container factory - we would need to return a different spec based on the type (or don't allow just the concurrency to be configured that way).

So, we could return a spec that supports all properties exception concurrency, or one that just configures `id`. 

WDYT?

Reply:
OK. I see what you mean. Never mind then.

Let's come back to your original idea about `getId() == null ? null : getId() + ".container"`!

We have similar convention for inner beans in other places, therefore this one should not hurt.
We should mention it in JavaDocs on the `KafkaMessageDrivenChannelAdapterSpec` since we don't have a good Docs for this project to explain there. 

Also need to mention that this id is going to be used only if there is no the `MessageListenerContainer` bean in the application context.

Although it might be good to reconsider that and register an `alias` with the provided non-generated id but that's already a SI Core story...